libvast/src/segment.cpp

```diff
@@ -0,0 +1,201 @@
+      caf::charbuf buf{payload + slice.start, slice_size};
+      caf::stream_deserializer<caf::charbuf&> deserializer{actor_system_, buf};
+      const_table_slice_handle x;
+      if (auto error = deserializer(x))
+        return error;
+      result.push_back(x);
+      // Fast forward to the ID one past this slice.
+      rng.skip(slice.offset + slice.size);
+    }
+  }
+  return result;
+}
+
+segment::segment(caf::actor_system& sys, chunk_ptr chunk) 
+  : actor_system_{sys},
+    chunk_{std::move(chunk)} {
+  // Only the builder and make() call this constructor. In the former case the
+  // chunk is properly constructed and in the latter case the factory ensures
+  // that it meets the requirements.
+  VAST_ASSERT(chunk_ != nullptr);
+}
+
+segment_builder::segment_builder(caf::actor_system& sys)
+  : actor_system_{sys},
+    table_slice_streambuf_{table_slice_buffer_},
+    table_slice_serializer_{actor_system_, table_slice_streambuf_} {
+  reset();
+}
+
+caf::error segment_builder::add(const_table_slice_handle x) {
+  if (x->offset() < min_table_slice_offset_)
+    return make_error(ec::unspecified, "table slice offsets not increasing");
+  auto before = table_slice_buffer_.size();
+  if (auto error = table_slice_serializer_(x)) {
+    table_slice_buffer_.resize(before);
+    return error;
+  }
+  auto after = table_slice_buffer_.size();
+  VAST_ASSERT(before < after);
+  meta_.slices.push_back({
+    detail::narrow_cast<int64_t>(before),
+    detail::narrow_cast<int64_t>(after),
+    x->offset(), x->rows()});
+  min_table_slice_offset_ = x->offset() + x->rows();
+  return caf::none;
+}
+
+caf::expected<segment_ptr> segment_builder::finish() {
+  auto guard = caf::detail::make_scope_guard([&] { reset(); });
+  // Write header.
+  segment_buffer_.resize(sizeof(segment::header));
+  auto header = reinterpret_cast<segment::header*>(segment_buffer_.data());
+  header->magic = to_little_endian(segment::magic);
+  header->version = to_little_endian(segment::version);
+  header->id = uuid::random();
+  // Serialize meta data into buffer.
+  caf::vectorbuf segment_streambuf{segment_buffer_};
+  detail::coded_serializer<caf::vectorbuf&> meta_serializer{segment_streambuf};
+  if (auto error = meta_serializer(meta_))
+    return error;
+  // Add the payload offset to the header.
+  header = reinterpret_cast<segment::header*>(segment_buffer_.data());
+  uint64_t payload_offset = segment_buffer_.size();
+  header->payload_offset = to_little_endian(payload_offset);
+  // Write the table slices.
+  segment_buffer_.resize(segment_buffer_.size() + table_slice_buffer_.size());
+  std::memcpy(segment_buffer_.data() + payload_offset,
+              table_slice_buffer_.data(),
+              table_slice_buffer_.size());
+  // Move the complete segment buffer into a chunk.
+  auto buffer = std::make_shared<std::vector<char>>(std::move(segment_buffer_));
+  auto deleter = [buf=buffer](char*, size_t) mutable { buf.reset(); };
```

Comment:
Maybe we could at least insert an assert that `ptr == buf.data()`?

Reply:
What do you mean by `ptr`?

Reply:
The currently unnamed `const char*` argument.

Reply:
Ah, okay. Such an assertion also strikes me as reasonable.